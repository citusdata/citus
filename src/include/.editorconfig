# Don't change generated input file
[citus_config.h.in]
insert_final_newline = unset
trim_trailing_whitespace = unset
indent_style = unset
indent_size = unset
