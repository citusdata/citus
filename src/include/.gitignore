/stamp-h
/stamp-ext-h
/citus_config.h
/citus_config.h.in~
/citus_version.h
/citus_version.h.in~
