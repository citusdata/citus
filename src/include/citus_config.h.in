/* src/include/citus_config.h.in.  Generated from configure.in by autoheader.  */


/*
 * citus_config.h.in is generated by autoconf/autoheader and
 * converted into citus_config.h by configure.  Include when code needs to
 * depend on determinations made by configure.
 *
 * Do not manually edit!
 */


/* Citus edition as a string */
#undef CITUS_EDITION

/* Extension version expected by this Citus build */
#undef CITUS_EXTENSIONVERSION

/* Citus major version as a string */
#undef CITUS_MAJORVERSION

/* Citus full name as a string */
#undef CITUS_NAME

/* Citus version as a string */
#undef CITUS_VERSION

/* Citus version as a number */
#undef CITUS_VERSION_NUM

/* A string containing the version number, platform, and C compiler */
#undef CITUS_VERSION_STR

/* Define to 1 if you have the <inttypes.h> header file. */
#undef HAVE_INTTYPES_H

/* Define to 1 if you have the <memory.h> header file. */
#undef HAVE_MEMORY_H

/* Define to 1 if you have the <stdint.h> header file. */
#undef HAVE_STDINT_H

/* Define to 1 if you have the <stdlib.h> header file. */
#undef HAVE_STDLIB_H

/* Define to 1 if you have the <strings.h> header file. */
#undef HAVE_STRINGS_H

/* Define to 1 if you have the <string.h> header file. */
#undef HAVE_STRING_H

/* Define to 1 if you have the <sys/stat.h> header file. */
#undef HAVE_SYS_STAT_H

/* Define to 1 if you have the <sys/types.h> header file. */
#undef HAVE_SYS_TYPES_H

/* Define to 1 if you have the <unistd.h> header file. */
#undef HAVE_UNISTD_H

/* Define to the address where bug reports for this package should be sent. */
#undef PACKAGE_BUGREPORT

/* Define to the full name of this package. */
#undef PACKAGE_NAME

/* Define to the full name and version of this package. */
#undef PACKAGE_STRING

/* Define to the one symbol short name of this package. */
#undef PACKAGE_TARNAME

/* Define to the home page for this package. */
#undef PACKAGE_URL

/* Define to the version of this package. */
#undef PACKAGE_VERSION

/* Base URL for statistics collection and update checks */
#undef REPORTS_BASE_URL

/* The size of `void *', as computed by sizeof. */
#undef SIZEOF_VOID_P

/* Define to 1 if you have the ANSI C header files. */
#undef STDC_HEADERS
