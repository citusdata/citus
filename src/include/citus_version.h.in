/* This file is created manually */

/* Citus full name as a string */
#undef CITUS_NAME

/* Citus edition as a string */
#undef CITUS_EDITION

/* Extension version expected by this Citus build */
#undef CITUS_EXTENSIONVERSION

/* Citus major version as a string */
#undef CITUS_MAJORVERSION

/* Citus version as a string */
#undef CITUS_VERSION

/* Citus version as a number */
#undef CITUS_VERSION_NUM

/* A string containing the version number, platform, and C compiler */
#undef CITUS_VERSION_STR

/* Define to 1 if you have the `curl' library (-lcurl). */
#undef HAVE_LIBCURL

/* Define to 1 if you have the `liblz4' library (-llz4). */
#undef HAVE_LIBLZ4

/* Define to 1 if you have the `libzstd' library (-lzstd). */
#undef HAVE_LIBZSTD

/* Base URL for statistics collection and update checks */
#undef REPORTS_BASE_URL

/* columnar table access method capability */
#undef HAS_TABLEAM
