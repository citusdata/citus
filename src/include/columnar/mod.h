/*-------------------------------------------------------------------------
 *
 * mod.h
 *
 * Type and function declarations for columnar
 *
 * Copyright (c) 2016, Citus Data, Inc.
 *
 * $Id$
 *
 *-------------------------------------------------------------------------
 */

#ifndef MOD_H
#define MOD_H

/* Function declarations for extension loading and unloading */
extern void columnar_init(void);
extern void columnar_fini(void);

#endif /* MOD_H */
