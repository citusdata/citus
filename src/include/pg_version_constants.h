/*-------------------------------------------------------------------------
 *
 * pg_version_constants.h
 *	   pg version related constants.
 *
 * Copyright (c) Citus Data, Inc.
 *
 *-------------------------------------------------------------------------
 */

#ifndef PG_VERSION_CONSTANTS
#define PG_VERSION_CONSTANTS

#define PG_VERSION_15 150000
#define PG_VERSION_16 160000
#define PG_VERSION_17 170000
#define PG_VERSION_18 180000

#endif   /* PG_VERSION_CONSTANTS */
