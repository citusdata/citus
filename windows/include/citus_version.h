/* Citus full name as a string */
#define CITUS_NAME "Citus"

/* Citus edition as a string */
#define CITUS_EDITION "community"

/* Extension version expected by this Citus build */
#define CITUS_EXTENSIONVERSION "7.3-1"

/* Citus major version as a string */
#define CITUS_MAJORVERSION "7.3"

/* Citus version as a string */
#define CITUS_VERSION "7.3devel"

/* Citus version as a number */
#define CITUS_VERSION_NUM 70300

/* A string containing the version number, platform, and C compiler */
#define CITUS_VERSION_STR "Citus 7.3devel on x86_64-windows, compiled by Visual Studio"

/* Define to 1 if you have the `curl' library (-lcurl). */
#define HAVE_LIBCURL 0

/* Base URL for statistics collection and update checks */
#define REPORTS_BASE_URL "https://reports.citusdata.com"
