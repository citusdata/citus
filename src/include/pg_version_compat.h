/*-------------------------------------------------------------------------
 *
 * pg_version_compat.h
 *	  Compatibility macros for writing code agnostic to PostgreSQL versions
 *
 * Copyright (c) Citus Data, Inc.
 *
 *-------------------------------------------------------------------------
 */

#ifndef PG_VERSION_COMPAT_H
#define PG_VERSION_COMPAT_H

#include "distributed/pg_version_constants.h"

#if PG_VERSION_NUM >= PG_VERSION_15
#define ProcessCompletedNotifies()
#define RelationCreateStorage_compat(a, b, c) RelationCreateStorage(a, b, c)
#define parse_analyze_varparams_compat(a, b, c, d, e) parse_analyze_varparams(a, b, c, d, \
																			  e)
#define CREATE_SEQUENCE_COMMAND \
	"CREATE %sSEQUENCE IF NOT EXISTS %s AS %s INCREMENT BY " INT64_FORMAT \
	" MINVALUE " INT64_FORMAT " MAXVALUE " INT64_FORMAT \
	" START WITH " INT64_FORMAT " CACHE " INT64_FORMAT " %sCYCLE"
#else

#include "nodes/value.h"
#include "storage/smgr.h"
#include "utils/int8.h"
#include "utils/rel.h"

typedef Value String;

#ifdef HAVE_LONG_INT_64
#define strtoi64(str, endptr, base) ((int64) strtol(str, endptr, base))
#define strtou64(str, endptr, base) ((uint64) strtoul(str, endptr, base))
#else
#define strtoi64(str, endptr, base) ((int64) strtoll(str, endptr, base))
#define strtou64(str, endptr, base) ((uint64) strtoull(str, endptr, base))
#endif
#define RelationCreateStorage_compat(a, b, c) RelationCreateStorage(a, b)
#define parse_analyze_varparams_compat(a, b, c, d, e) parse_analyze_varparams(a, b, c, d)
#define pgstat_init_relation(r) pgstat_initstats(r)
#define pg_analyze_and_rewrite_fixedparams(a, b, c, d, e) pg_analyze_and_rewrite(a, b, c, \
																				 d, e)
#define boolVal(v) intVal(v)
#define makeBoolean(val) makeInteger(val)

static inline int64
pg_strtoint64(char *s)
{
	int64 result;
	(void) scanint8(s, false, &result);
	return result;
}


static inline SMgrRelation
RelationGetSmgr(Relation rel)
{
	if (unlikely(rel->rd_smgr == NULL))
	{
		smgrsetowner(&(rel->rd_smgr), smgropen(rel->rd_node, rel->rd_backend));
	}
	return rel->rd_smgr;
}


#define CREATE_SEQUENCE_COMMAND \
	"CREATE SEQUENCE IF NOT EXISTS %s AS %s INCREMENT BY " INT64_FORMAT \
	" MINVALUE " INT64_FORMAT " MAXVALUE " INT64_FORMAT \
	" START WITH " INT64_FORMAT " CACHE " INT64_FORMAT " %sCYCLE"

#endif

#if PG_VERSION_NUM >= PG_VERSION_14
#define AlterTableStmtObjType_compat(a) ((a)->objtype)
#define getObjectTypeDescription_compat(a, b) getObjectTypeDescription(a, b)
#define getObjectIdentity_compat(a, b) getObjectIdentity(a, b)

/* for MemoryContextMethods->stats */
#define stats_compat(a, b, c, d, e) stats(a, b, c, d, e)
#define FuncnameGetCandidates_compat(a, b, c, d, e, f, g) \
	FuncnameGetCandidates(a, b, c, d, e, f, g)
#define expand_function_arguments_compat(a, b, c, d) expand_function_arguments(a, b, c, d)
#define BeginCopyFrom_compat(a, b, c, d, e, f, g, h) BeginCopyFrom(a, b, c, d, e, f, g, h)
#define standard_ProcessUtility_compat(a, b, c, d, e, f, g, h) \
	standard_ProcessUtility(a, b, c, d, e, f, g, h)
#define ProcessUtility_compat(a, b, c, d, e, f, g, h) \
	ProcessUtility(a, b, c, d, e, f, g, h)
#define PrevProcessUtility_compat(a, b, c, d, e, f, g, h) \
	PrevProcessUtility(a, b, c, d, e, f, g, h)
#define SetTuplestoreDestReceiverParams_compat(a, b, c, d, e, f) \
	SetTuplestoreDestReceiverParams(a, b, c, d, e, f)
#define pgproc_statusflags_compat(pgproc) ((pgproc)->statusFlags)
#define get_partition_parent_compat(a, b) get_partition_parent(a, b)
#define RelationGetPartitionDesc_compat(a, b) RelationGetPartitionDesc(a, b)
#define make_simple_restrictinfo_compat(a, b) make_simple_restrictinfo(a, b)
#define pull_varnos_compat(a, b) pull_varnos(a, b)
#else
#define AlterTableStmtObjType_compat(a) ((a)->relkind)
#define F_NEXTVAL F_NEXTVAL_OID
#define ROLE_PG_MONITOR DEFAULT_ROLE_MONITOR
#define PROC_WAIT_STATUS_WAITING STATUS_WAITING
#define getObjectTypeDescription_compat(a, b) getObjectTypeDescription(a)
#define getObjectIdentity_compat(a, b) getObjectIdentity(a)

/* for MemoryContextMethods->stats */
#define stats_compat(a, b, c, d, e) stats(a, b, c, d)
#define FuncnameGetCandidates_compat(a, b, c, d, e, f, g) \
	FuncnameGetCandidates(a, b, c, d, e, g)
#define expand_function_arguments_compat(a, b, c, d) expand_function_arguments(a, c, d)
#define VacOptValue VacOptTernaryValue
#define VACOPTVALUE_UNSPECIFIED VACOPT_TERNARY_DEFAULT
#define VACOPTVALUE_DISABLED VACOPT_TERNARY_DISABLED
#define VACOPTVALUE_ENABLED VACOPT_TERNARY_ENABLED
#define CopyFromState CopyState
#define BeginCopyFrom_compat(a, b, c, d, e, f, g, h) BeginCopyFrom(a, b, d, e, f, g, h)
#define standard_ProcessUtility_compat(a, b, c, d, e, f, g, h) \
	standard_ProcessUtility(a, b, d, e, f, g, h)
#define ProcessUtility_compat(a, b, c, d, e, f, g, h) ProcessUtility(a, b, d, e, f, g, h)
#define PrevProcessUtility_compat(a, b, c, d, e, f, g, h) \
	PrevProcessUtility(a, b, d, e, f, g, h)
#define COPY_FRONTEND COPY_NEW_FE
#define SetTuplestoreDestReceiverParams_compat(a, b, c, d, e, f) \
	SetTuplestoreDestReceiverParams(a, b, c, d)
#define pgproc_statusflags_compat(pgproc) \
	((&ProcGlobal->allPgXact[(pgproc)->pgprocno])->vacuumFlags)
#define get_partition_parent_compat(a, b) get_partition_parent(a)
#define RelationGetPartitionDesc_compat(a, b) RelationGetPartitionDesc(a)
#define PQ_LARGE_MESSAGE_LIMIT 0
#define make_simple_restrictinfo_compat(a, b) make_simple_restrictinfo(b)
#define pull_varnos_compat(a, b) pull_varnos(b)
#define ROLE_PG_READ_ALL_STATS DEFAULT_ROLE_READ_ALL_STATS
#endif

#define SetListCellPtr(a, b) ((a)->ptr_value = (b))
#define RangeTableEntryFromNSItem(a) ((a)->p_rte)
#define fcGetArgValue(fc, n) ((fc)->args[n].value)
#define fcGetArgNull(fc, n) ((fc)->args[n].isnull)
#define fcSetArgExt(fc, n, val, is_null) \
	(((fc)->args[n].isnull = (is_null)), ((fc)->args[n].value = (val)))
#define fcSetArg(fc, n, value) fcSetArgExt(fc, n, value, false)
#define fcSetArgNull(fc, n) fcSetArgExt(fc, n, (Datum) 0, true)

#endif   /* PG_VERSION_COMPAT_H */
