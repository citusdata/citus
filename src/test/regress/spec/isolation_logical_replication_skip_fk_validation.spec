#include "isolation_mx_common.include.spec"

setup
{
  SET citus.shard_count to 1;
  SET citus.shard_replication_factor to 1;

  CREATE TABLE t1 (id int PRIMARY KEY, value int);
  SELECT create_distributed_table('t1', 'id');

  CREATE TABLE t2 (id int PRIMARY KEY, value int);
  SELECT create_distributed_table('t2', 'id');

  CREATE TABLE r (id int PRIMARY KEY, value int);
  SELECT create_reference_table('r');
  SELECT get_shard_id_for_distribution_column('t1', 5) INTO selected_shard_for_test_table;
}

setup {
  ALTER TABLE t1 ADD CONSTRAINT t1_t2_fkey FOREIGN KEY (id) REFERENCES t2(id);
}

setup {
  ALTER TABLE t1 ADD CONSTRAINT t1_r_fkey FOREIGN KEY (value) REFERENCES r(id);
}

teardown
{
  DROP TABLE t1;
  DROP TABLE t2;
  DROP TABLE r;
  DROP TABLE selected_shard_for_test_table;
}

session "s1"

step "s1-start-session-level-connection"
{
	SELECT start_session_level_connection_to_node('localhost', 57638);
}

// This inserts a foreign key violation directly into the shard on the target
// worker. Since we're not validating the foreign key on the new shard on
// purpose we expect no errors.
step "s1-insert-violation-into-shard"
{
	SELECT run_commands_on_session_level_connection_to_node(format('INSERT INTO t1_%s VALUES (-1, -1)', (SELECT * FROM selected_shard_for_test_table)));
}

session "s2"

step "s2-move-placement"
{
	SELECT master_move_shard_placement((SELECT * FROM selected_shard_for_test_table), 'localhost', 57637, 'localhost', 57638);
}

session "s3"

// this advisory lock with (almost) random values are only used
// for testing purposes. For details, check Citus' logical replication
// source code
step "s3-acquire-advisory-lock"
{
	SELECT pg_advisory_lock(44000, 55152);
}

step "s3-release-advisory-lock"
{
	SELECT pg_advisory_unlock(44000, 55152);
}

permutation "s1-start-session-level-connection" "s3-acquire-advisory-lock" "s2-move-placement" "s1-start-session-level-connection" "s1-insert-violation-into-shard" "s3-release-advisory-lock"
