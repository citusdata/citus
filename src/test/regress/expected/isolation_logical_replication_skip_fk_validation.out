Parsed test spec with 3 sessions

starting permutation: s1-start-session-level-connection s3-acquire-advisory-lock s2-move-placement s1-start-session-level-connection s1-insert-violation-into-shard s3-release-advisory-lock
step s1-start-session-level-connection:
 SELECT start_session_level_connection_to_node('localhost', 57638);

start_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step s3-acquire-advisory-lock:
 SELECT pg_advisory_lock(44000, 55152);

pg_advisory_lock
---------------------------------------------------------------------

(1 row)

step s2-move-placement:
 SELECT master_move_shard_placement((SELECT * FROM selected_shard_for_test_table), 'localhost', 57637, 'localhost', 57638);
 <waiting ...>
step s1-start-session-level-connection: 
 SELECT start_session_level_connection_to_node('localhost', 57638);

start_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step s1-insert-violation-into-shard:
 SELECT run_commands_on_session_level_connection_to_node(format('INSERT INTO t1_%s VALUES (-1, -1)', (SELECT * FROM selected_shard_for_test_table)));

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step s3-release-advisory-lock:
 SELECT pg_advisory_unlock(44000, 55152);

pg_advisory_unlock
---------------------------------------------------------------------
t
(1 row)

step s2-move-placement: <... completed>
master_move_shard_placement
---------------------------------------------------------------------

(1 row)

